### PR TITLE
Avoid `Windows.h`

### LINKED ISSUE #599: 6$: Avoid <Windows.h>
#include more specialized headers instead of <Windows.h>: https://aras-p.info/blog/2018/01/12/Minimizing-windows.h/

Goal: reduce compile times.

### PR DESCRIPTION
**Pull request by EmptyGroceryBag**
It turns out that QLJS uses a vanishingly small portion of the win32
API. Therefore, it's unnecessary to include `Windows.h` and all the
bloat it drags in with it. Modularize include directives to only include
certain portions of the win32 API.

This commit should not change behavior.

Fixes #599

### CODE REVIEW

**Comment by EmptyGroceryBag** [comment]
Note: I just commented out the `#include <Windows.h>` lines for now. If you approve of my changes, I'll go ahead and clean it up tomorrow.

**Comment by EmptyGroceryBag** [comment]
I have read and hereby agree to quick-lint-js' Contributor License Agreement (CLA-v1.md).

File: src/change-detecting-filesystem-win32.cpp [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 #if defined(_WIN32)
 
-#include <Windows.h>
+//#include <Windows.h>
```
Comment (strager) [inline src/change-detecting-filesystem-win32.cpp:6]: Doesn't this file use Windows APIs? We should `#include` the headers for the functions and classes this module uses.

File: src/file-path.cpp [hunk 2]
```diff
@@ -8,8 +8,31 @@
...
+#endif
+
+#if !defined(_68K_) && !defined(_MPPC_) && !defined(_X86_) && !defined(_IA64_) && !defined(_AMD64_) && !defined(_ARM_) && !defined(_ARM64_) && defined(_M_ARM)
+#define _ARM_
+#endif
+
+#if !defined(_68K_) && !defined(_MPPC_) && !defined(_X86_) && !defined(_IA64_) && !defined(_AMD64_) && !defined(_ARM_) && !defined(_ARM64_) && defined(_M_ARM64)
+#define _ARM64_
+#endif
+// ------
```
Comment (strager) [inline src/file-path.cpp:32]: Must fix: Do not include Microsoft-copyright unlicensed code in our repository.

File: src/file.cpp [hunk 3]
```diff
@@ -35,7 +35,14 @@
...
 
 #if QLJS_HAVE_WINDOWS_H
+/*
 #include <Windows.h>
+#include <winerror.h>
+#include <errhandlingapi.h>
+#include <processenv.h>
+#include <fileapi.h>
+*/
+#include <WinBase.h>
```
Comment (strager) [inline src/file.cpp:45]: Same comment as on src/change-detecting-filesystem-win32.cpp.

File: src/utf-16.cpp [hunk 4]
```diff
@@ -14,6 +14,7 @@
 
 #if defined(_WIN32)
 #include <Windows.h>
```
Comment (strager) [inline src/utf-16.cpp:16]: You missed a spot.

File: test/test-file-canonical.cpp [hunk 5]
```diff
@@ -26,6 +26,10 @@
 #include <quick-lint-js/string-view.h>
 #include <string>
 
+#if QLJS_HAVE_WINDOWS_H
+#include <winerror.h>
+#endif
```
Comment (strager) [inline test/test-file-canonical.cpp:31]: :+1:

**Review by strager — CHANGES_REQUESTED** [review]

File: src/change-detecting-filesystem-win32.cpp (hunk 1 above)
Comment (EmptyGroceryBag) [reply src/change-detecting-filesystem-win32.cpp:6]: This file includes other files that include the necessary Windows APIs. Is it good form to explicitly include APIs that are used by this file?

File: src/file-path.cpp (hunk 2 above)
Comment (EmptyGroceryBag) [reply src/file-path.cpp:32]: `Windows.h` seems to be unavoidable here.

File: src/change-detecting-filesystem-win32.cpp (hunk 1 above)
Comment (strager) [reply src/change-detecting-filesystem-win32.cpp:6]: > Is it good form to explicitly include APIs that are used by this file?

Yes. [Include what you use](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md).

**Comment by EmptyGroceryBag** [comment]
It took me a few passes, but I eventually figured out that most of the time all that is needed is `minwindef.h`, which is only 286 lines.

File: src/file.cpp [hunk 6]
```diff
@@ -2,7 +2,6 @@
 // See end of file for extended copyright information.
 
 #if !defined(__EMSCRIPTEN__)
-
```
Comment (strager) [inline src/file.cpp:5]: Nit: undo

File: src/file.cpp [hunk 7]
```diff
@@ -35,7 +34,8 @@
 #endif
 
 #if QLJS_HAVE_WINDOWS_H
-#include <Windows.h>
+//#include <Windows.h>
+#include <minwindef.h>
```
Comment (strager) [inline src/file.cpp:38]: Hmm, does this really declare CreateFile?

**Comment by EmptyGroceryBag** [comment]
Rip. `clang-format` changed the order of the `#include` directives and broke the build. Revising, _again_.

File: src/quick-lint-js/windows.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+#define OEMRESOURCE
+#define NOATOM
+#define NOCLIPBOARD
+#define NOCOLOR
+#define NOCTLMGR
+#define NODRAWTEXT
+#define NOGDI
+#define NOKERNEL
+#define NOUSER
+//#define NONLS
```
Comment (strager) [inline src/quick-lint-js/windows.h:32]: Nit: Remove

File: src/quick-lint-js/windows.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+#define NOSOUND
+#define NOTEXTMETRIC
+#define NOWH
+#define NOWINOFFSETS
+#define NOCOMM
+#define NOKANJI
+#define NOHELP
+#define NOPROFILER
+#define NODEFERWINDOWPOS
+#define NOMCX
```
Comment (strager) [inline src/quick-lint-js/windows.h:50]: Nit: Sort this list alphabetically.

**Review by strager — APPROVED** [review]
Looks good. You should remove the commented-out `#include`-s.

How much does this patch speed up compilation?

**Review by strager — APPROVED** [review]

File: .clang-format [hunk 10]
```diff
@@ -4,7 +4,10 @@
 BasedOnStyle: Google
 
 IncludeBlocks: Merge
-IncludeCategories: [{Regex: '.*', Priority: 0}]
+IncludeCategories: [
+	{Regex: '<pathcch\.h>', Priority: 1},
```
Comment (strager) [inline .clang-format:8]: Tabs! 😱

**Review by strager — APPROVED** [review]

**Pull request merged**